File: TP1/11_palindromo_recursivo/Palindromo_recursivo.java
/**
 * Marcos Antonio Lommez Candido Ribeiro
 * 77157
 * AED's 2
 * 10/08/2022
 */

public class Palindromo_recursivo {

	/**
	 * Metodo para testar se a String contem a palavra 'FIM'
	 * 
	 * @param s String a testar
	 * @return True se for 'FIM'
	 */
	public static boolean isFim(String s) {
		return (s.length() == 3 && s.charAt(0) == 'F' && s.charAt(1) == 'I' && s.charAt(2) == 'M');
	}

	/**
	 * Metodo inicializador do metodo de recursivo
	 * @param s String a ser testada
	 * @return True se for um palindromo
	 */
	public static boolean ePalindromo(String s) {
		return ePalindromo(s, 0, (s.length() - 1));
	}

	/**
	 * Metodo recursivo que recebe uma String e testa se
	 * a mesma e' um palindromo, exemplo:
	 * arara, reviver, ama, ovo
	 * 
	 * @param s String a ser testada
	 * @param i Inicio da String
	 * @param j Final da String
	 * @return True se for um palindromo
	 */
	public static boolean ePalindromo(String s, int i, int j) {
		boolean resp;
		if (i >= j) {
			resp = true;
		} else if (s.charAt(i) != s.charAt(j)) {
			resp = false;
		} else {
			resp = ePalindromo(s, i + 1, j - 1);
		}
		return resp;
	}

	// Driver
	public static void main(String[] args) {
		String[] entrada = new String[1000];
		int numEntrada = 0;

		// Leitura da entrada padrao
		do {
			entrada[numEntrada] = MyIO.readLine();
		} while (isFim(entrada[numEntrada++]) == false);
		numEntrada--; // Desconsiderar ultima linha contendo a palavra FIM

		for (int i = 0; i < numEntrada; i++) {
			System.out.println(ePalindromo(entrada[i]) ? "SIM" : "NAO");
		}
	}
}

File: TP1/12_palindromo_C/Palindromo_recursivo.c
/**
 * Marcos Antonio Lommez Candido Ribeiro
 * 77157
 * AED's 2
 * 21/08/2022
 */

#include <stdio.h>
#include <stdbool.h>
#include <string.h>

/**
 * Metodo recursivo que recebe uma String e testa se
 * a mesma e' um palindromo, exemplo:
 * arara, reviver, ama, ovo
 *
 * @param s String a ser testada
 * @param i Inicio da String
 * @param j Final da String
 * @return True se for um palindromo
 */
bool ePalindromoRec(char s[], int i, int j)
{
    bool resp = true;
    if (i >= j)
    {
        resp = true;
    }
    else if (s[i] != s[j])
    {
        resp = false;
    }
    else
    {
        resp = ePalindromoRec(s, i + 1, j - 1);
    }
    return resp;
}

/**
 * Metodo inicializador do metodo de recursivo
 * @param s String a ser testada
 * @return True se for um palindromo
 */
bool ePalindromo(char s[])
{
    return ePalindromoRec(s, 0, (strlen(s) - 1));
}

/**
 * Metodo para testar se a String contem a palavra 'FIM'
 *
 * @param s String a testar
 * @return True se for 'FIM'
 */
bool isFim(char s[])
{
    return (strlen(s) == 3 && s[0] == 'F' && s[1] == 'I' && s[2] == 'M');
}

// Driver
int main()
{
    char entrada[1000];

    for (; !isFim(entrada);)
    {
        scanf(" %1000[^\n]", entrada);
        if (!isFim(entrada))
            printf("%s\n", ePalindromo(entrada) ? "SIM" : "NAO");
    }

    return 0;
}
